### DISCUSSION TITLE
Javascript Hook not working

### DISCUSSION THREAD

**Discussion by badcom** [livewire/livewire#7419] [7. Bugs (archived)]
### Livewire version

3.2.2

### Which PHP version are you using?

PHP 8.2

### Repository URL

https://github.com/badcom/livewire-bug

### Steps To Reproduce

1. Access the home page
2. Click the "Click me!" button
3. A message should get sent to the console via the Livewire.hook('commit') hook, but it's never sent.

### Additional information

Take a look at the test.blade.php file. There's a Livewire.hook code at the bottom which should send a message to the console, but the message never gets sent. Am I missing something?
```
@script
<script>
    Livewire.hook('commit', ({ component, commit, respond, succeed, fail }) => {
        console.log('commited!')
    })
</script>
@endscript
```

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by lrljoe** [comment]
I suspect this is being called prior to Livewire actually being present, you can confirm this by using your Browser Console and seeing if there's an error "Livewire is not defined".  In which case, can you please try wrapping your method inside:
```js
        document.addEventListener('livewire:initialized', () => {

        });
```

Now, interestingly:
```
@script
<script>
    alert('First Load');
        Livewire.hook('commit.prepare', ({ component, commit }) => {
            alert('committing234');
        });
</script>
@endscript
```
Will fire a "First Load" alert when the page renders, and will then fire an alert "committing234" on subsequent actions.

Whereas
```
@script
<script>
        Livewire.hook('commit.prepare', ({ component, commit }) => {
            alert('committing234');
        });
</script>
@endscript
```
Will just not behave at all.

It's almost as if it doesn't like Livewire until the page has had some other JS activity. Will do some digging if nobody else fixes it in the next week or so.

**Comment by phred** [comment]
I don't know if this is the same issue but I can't seem get `Livewire.hook` to fire at all. I found an unrelated bug (using `@script` inside a lazy-loaded component triggers a `using $this not in object context` error server-side), so I moved my code to my app.js:

```
  document.addEventListener("livewire:initialized", () => {
    console.log("livewire init");
    Livewire.hook("component.init", ({ component, cleanup }) => {
      console.log("component init", component);
      if (component.name == "cart-count") {
        console.log("hi");
      }
    });
  );
```

I only see `livewire init` in my browser, and the `component.init` hook is never called. Any idea what's going on @lrljoe ?

**Reply by lrljoe** [reply]
Where is that code sitting in terms of your application?

If you add this to your Layout (in the head) - do you get a "hi" when you have cart-count component loaded?

```
document.addEventListener('livewire:init', () => {
    console.log("livewire init");
    Livewire.hook('component.init', ({ component, cleanup }) => {
        console.log("component init", component);
        if (component.name == "cart-count") {
        console.log("hi");
      }
    });
  });
```

Note - you want to use init, not initialized in this use case.

**Reply by phred** [reply]
My `app.js` is compiled by Vite and injected in the head of my layout.  Here's the relevant part of my app.js:

```
if (typeof document !== "undefined") {
  let _paymentComplete = false;
  const finalElement = "checkout-final";

  document.addEventListener("livewire:initialized", () => {
    console.log("livewire init");
    Livewire.hook("component.init", ({ component, cleanup }) => {
      console.log("component init", component);
      if (component.name == "cart-count") {
        console.log("hi");
      }
    });

    Livewire.on("needs-payment-info", () => {
      scrollToFinalIfNeeded();
    });
  });
// ...
```

The "needs-payment-info" event handling works great, though it's unrelated to `cart-count`, it's sent by a different component on the page.

Ah, yes! @lrljoe the "component.init" does fire if I put your code in a `<script>` tag in the head of my layout. I'd prefer not to do that though. Why does it make a difference?

**Reply by EdgarsJoja** [reply]
In docs it is mentioned, that:
> It is often beneficial to register any [custom directives](https://livewire.laravel.com/docs/javascript#registering-custom-directives) or [lifecycle hooks](https://livewire.laravel.com/docs/javascript#javascript-hooks) inside of livewire:init so that they are available before Livewire begins initializing on the page.

Might be that "initialized" event is too late, and other hook events have already been dispatched.

**Reply by lrljoe** [reply]
Try with init, not initialized in your app.js

You probably want to register your hooks as Livewire is initialising, not after it has.

**Comment by phred** [comment]
Ah yep, changing to `document.addEventListener("livewire:init", () => {})` in my app.js made all the difference. I guess `livewire.initialized` is from LiveWire 2: https://livewire.laravel.com/docs/upgrading#new-lifecycle-hooks

Interesting that it still fires, albeit at a different point in the component setup. Might be a good candidate for a console.warn(), perhaps.

I am now less sure that my issue was related to the original issue. Thanks for the help @EdgarsJoja and @lrljoe.
